### PR TITLE
Migrate usage of checkData() and validateDataExists() to validateData()

### PR DESCRIPTION
**Pull request by lannka**
This PR introduced new function `validateData(data, mandatoryFields, optionalFields)`. 

`checkData(data, allowedFields)` is replaced by `validateData(data, [ ], allowedFields`);
`validateDataExists(data, mandatoryFields)` is replaced by `validateData(data, mandatoryFields)`;

This would potentially encourage new code to check mandatory fields. See rationale in #4598

### CODE REVIEW

File: 3p/3p.js [hunk 1]
```diff
@@ -211,15 +194,34 @@ export function computeInMasterFrame(global, taskId, work, cb) {
...
-export function validateDataExists(data, mandatoryFields) {
+export function validateData(data, mandatoryFields, opt_optionalFields) {
+  let allowedFields = opt_optionalFields || [];
   for (let i = 0; i < mandatoryFields.length; i++) {
     const field = mandatoryFields[i];
-    user().assert(data[field],
-        'Missing attribute for %s: %s.', data.type, field);
+    if (Array.isArray(field)) {
+      validateExactlyOne(data, field);
+      allowedFields = allowedFields.concat(field);
```
Comment (jridgewell) [inline 3p/3p.js]: Let's do this afterward the loop, instead of creating `n` arrays.

File: 3p/3p.js [hunk 2]
```diff
@@ -263,12 +265,20 @@ export function validateData(data, allowedFields) {
...
   };
-  for (const field in data) {
-    if (!data.hasOwnProperty(field) ||
-        field in defaultAvailableFields) {
-      continue;
+
+  // Throw in a timeout, because we do not want to interrupt execution,
+  // because that would make each removal an instant backward incompatible
+  // change.
+  try {
```
Comment (jridgewell) [inline 3p/3p.js]: There shouldn't be any need for the try-catch. We can just call `rethrowAsync` in the for loop and `break`.

File: 3p/3p.js (hunk 1 above)
Comment (jridgewell) [inline 3p/3p.js]: Why is this being concatenated?

Comment (lannka) [inline 3p/3p.js]: Here is an example of `mandatoryFields`:

```
[
  'foo', 'bar',
  ['hello', 'world']
]
```

This is to flatten the array to 1 level: 
```
['foo', 'bar', 'hello', 'world']
```

As they're all allowed fields.

File: 3p/3p.js (hunk 2 above)
Comment (lannka) [inline 3p/3p.js]: good point.

File: 3p/3p.js (hunk 1 above)
Comment (jridgewell) [inline 3p/3p.js]: We can avoid the multiple concats and pushes by just concatenating after the loop, then.

Comment (lannka) [inline 3p/3p.js]: Can you give an example?

Comment (jridgewell) [inline 3p/3p.js]:
```js
export function validateData(data, mandatoryFields, opt_optionalFields) {
  let allowedFields = opt_optionalFields || [];
  for (let i = 0; i < mandatoryFields.length; i++) {
    const field = mandatoryFields[i];
    if (Array.isArray(field)) {
      validateExactlyOne(data, field);
    } else {
      user().assert(data[field],
          'Missing attribute for %s: %s.', data.type, field);
    }
  }
  validateAllowedFields(data, mandatoryFields.concat(allowedFields));
}

Comment (lannka) [inline 3p/3p.js]: But I want flatten `mandatorFields`, it's an `Array<string|Array<string>>`

Comment (jridgewell) [inline 3p/3p.js]: Ohhhh, never mind.

Comment (zhouyx) [inline 3p/3p.js]: can we only push the one filed that satisfy data[field] to `allowedField`?

File: 3p/3p.js [hunk 3]
```diff
@@ -211,15 +194,34 @@ export function computeInMasterFrame(global, taskId, work, cb) {
...
+    if (Array.isArray(field)) {
+      validateExactlyOne(data, field);
+      allowedFields = allowedFields.concat(field);
+    } else {
+      user().assert(data[field],
+          'Missing attribute for %s: %s.', data.type, field);
+      allowedFields.push(field);
+    }
+  }
+  if (opt_optionalFields) {
```
Comment (zhouyx) [inline 3p/3p.js]: shouldn't this be `allowedFields`?

File: ads/adman.js [hunk 4]
```diff
@@ -14,19 +14,16 @@
...
  * @param {!Window} global
  * @param {!Object} data
  */
 export function adman(global, data) {
-  const script = global.document.createElement('script');
-  const fields = ['ws', 'host', 's'];
-
-  checkData(data, fields);
-  validateDataExists(data, fields);
+  validateData(data, ['ws', 'host', 's'], []);
```
Comment (zhouyx) [inline ads/adman.js]: `validateData(data, ['ws', 'host', 's']);` ?

File: ads/caprofitx.js [hunk 5]
```diff
@@ -14,14 +14,14 @@
...
+import {loadScript, validateData} from '../3p/3p';
+
 /**
  * @param {!Window} global
  * @param {!Object} data
  */
 export function caprofitx(global, data) {
-  checkData(data, ['tagid']);
-  validateDataExists(data, ['tagid']);
+  validateData(data, ['tagid'], []);
```
Comment (zhouyx) [inline ads/caprofitx.js]: `validateData(data, ['tagid']);`?

Comment (zhouyx) [inline ads/caprofitx.js]: Ah! I see this is for [] is added for `checkData()`. This is a bit confusing to me

File: 3p/3p.js (hunk 3 above)
Comment (jridgewell) [inline 3p/3p.js]: I think this check is actually useless, we should always run the code.

Comment (zhouyx) [inline 3p/3p.js]: agree

File: 3p/3p.js [hunk 6]
```diff
@@ -211,15 +194,34 @@ export function computeInMasterFrame(global, taskId, work, cb) {
...
+ * opt_optionalFields, it also validates that the data contains no fields other
+ * than mandatory and optional fields.
+ *
+ * Mandatory fields also accept a string Array as an item. All items in that
+ * array are considered as alternatives to each other. So the validation checks
+ * that the data contains exactly one of those alternatives.
+ *
  * @param {!Object} data
- * @param {!Array<string>} mandatoryFields
+ * @param {!Array<string|Array<string>>} mandatoryFields
```
Comment (erwinmombay) [inline 3p/3p.js]: {!Array<string|`!`Array<string>>} (missing the !)

File: 3p/3p.js (hunk 3 above)
Comment (lannka) [inline 3p/3p.js]: `validateData(data, mandatory)` allows arbitrary fields.

`validateDate(data, mandatory, optional)` allows fields of mandatory + optional.

File: ads/adman.js (hunk 4 above)
Comment (lannka) [inline ads/adman.js]: `validateData(data, ['ws', 'host', 's'], []);` means there's no `optionalFields`, so only `['ws', 'host', 's']` are allowed.

File: 3p/3p.js (hunk 6 above)
Comment (lannka) [inline 3p/3p.js]: nice catch!

File: 3p/3p.js (hunk 1 above)
Comment (lannka) [inline 3p/3p.js]: > can we only push the one filed that satisfy data[field] to allowedField?

nahh, that makes code confusing... and it doesn't really optimize too much.

File: 3p/3p.js (hunk 3 above)
Comment (zhouyx) [inline 3p/3p.js]: get your point! A bit confusing as optional = null and optional = [] behave very differently.

Comment (lannka) [inline 3p/3p.js]: we basically encourage users to use the more strict form:
`validateDate(data, mandatory, optional)`

but we have to be backward compatible, many existing 3Ps doesn't check optional fields.

Comment (zhouyx) [inline 3p/3p.js]: if user will eventually use `validateDate(data, mandatory, optional)`, i am good with this

**Comment by zhouyx** [comment]
LGTM

**Comment by lannka** [comment]
@jridgewell any more comments?

**Comment by jridgewell** [comment]
LGTM.

**Pull request merged**